config/pmd-main.xml

```diff
@@ -10,19 +10,10 @@
     <exclude-pattern>.*/src/it/.*</exclude-pattern>
     <exclude-pattern>.*/src/test/.*</exclude-pattern>
     <rule ref="config/pmd.xml"/>
-
-    <rule ref="rulesets/java/unusedcode.xml/UnusedPrivateMethod">
```

Comment:
I see removal of UnusedPrivateMethod from our configs only, rule  UnusedPrivateMethod is still present in pmd6
https://pmd.github.io/pmd-6.0.0/pmd_rules_java_bestpractices.html

please explain.

Reply:
You mean "removal of UnusedPrivateMethod _suppressions_", did you?
They were the false-positives and were fixed in PMD6.

Let me explain how I did this:
- First I deleted all the rules.
- Then I added every category entirely, one category by turn. For example:
```
<rule ref="category/java/bestpractices.xml"/>
```
This led to some violations. For each violation, I checked the old configuration for appropriate suppression. If the corresponding one was, I copied it. Otherwise, I added a new suppression (there were two of them, see #5638).
If a rule does not produce any violations, I did nothing for it. Thats why some suppressions are gone.

Reply:
> Otherwise, I added a new suppression (there were two of them,

please mark all such temporal suppression by comment "till https:....."
To let us clearly see what is new and temporal suppression

Reply:
Done:
```
<!-- Till https://github.com/checkstyle/checkstyle/issues/5638 -->
     <exclude name="AvoidFileStream"/>
```

For `OptimizableToArrayCall`: I'm going to leave it suppressed.
I run the benchmark from the [article](https://shipilev.net/blog/2016/arrays-wisdom-ancients/) and on my PC with 1.8.0_151 the fullsized version is faster when zero sized.

Reply:
if you decide to suppress it, please put in explanation link to this awesome article.

Reply:
I just remembered. There was one more: `DataClass` which complains on almost every class.
It suppressed with comment
    <!-- It produces too much violations, suppressed till we figure out how useful that metrics -->
(copied from `LawOfDemeter` rule suppression)

Also, some rules were renamed. Rule `NcssMethodCount` now just `NcssCount`

The full list is [here](https://pmd.github.io/pmd-6.0.0/pmd_release_notes.html#new-rules).

Some new rules are duplicate the CS checks, like the `EmptyCatchBlock` rule.
They should be suppressed too. Just for performance reason. I'll update the PR later.

Reply:
> They should be suppressed too. Just for performance reason. 

No good reason to suppress.
We suppress only if it produce too much false positives, or a lot of problems and there are analog in other places.

Non of library is ideal in implementation of some Rule (checkstyle,pmd, spotbugs,Sonar, Idea Inspections,........) all have problems, so twhen you use all of them, at least on of them will catch a problem.